interp/run.go

```diff
@@ -852,20 +864,37 @@ func callBin(n *node) {
 				for i, v := range values {
 					in[i] = v(f)
 				}
-				out := value(f).Call(in)
+				out, ok := callWithContext(ctx, value(f), in)
+				if !ok {
+					return nil
+				}
 				copy(f.data[n.findex:], out)
 				return tnext
 			}
 		}
 	}
 }
 
-func getIndexBinMethod(n *node) {
+func callWithContext(ctx context.Context, fn reflect.Value, in []reflect.Value) (out []reflect.Value, ok bool) {
```

Comment:
I think that `callWithContext` suffers from the same performance problem that was exposed in your previous change of `runCfg`: you add a goroutine+select for each call of a binary function!!! In addition, it's a kind of useless as we have no way to interrupt the binary call itself. If it is a matter of not stalling the interpreter as a whole, we could just spawn just one goroutine in the top `eval` call instead.

Reply:
>If it is a matter of not stalling the interpreter as a whole, we could just spawn just one goroutine in the top eval call instead.

I think this is a good solution. If we have the per exec check that's in the first commit for live-lock and a global goroutine/select for dead lock then the code changes are less than here and there is only one goroutine/select. I doubt that the performance impact it likely to be as bad a your are worried about, but this is a better solution.